python/pyxet/src/lib.rs

```diff
@@ -463,212 +522,417 @@ impl PyRFile {
+// errors elsewhere.
+//
 #[pyclass(subclass)]
-#[derive(Default)]
-struct PyWriteTransaction {
-    transaction: Option<XetRepoWriteTransaction>,
+#[derive(Clone)]
+pub struct PyWriteTransaction {
+    pwt: Option<Arc<RwLock<WriteTransaction>>>,
+
+    repo: Arc<XetRepo>,
     branch: String,
-    #[pyo3(get)]
-    new_files: Vec<String>,
-    #[pyo3(get)]
-    copies: Vec<(String, String)>,
-    #[pyo3(get)]
-    deletes: Vec<String>,
-    #[pyo3(get)]
-    moves: Vec<(String, String)>,
-    do_not_commit: bool,
-    error_on_commit: bool,
-    ready_to_commit: bool,
-    pending_write_files: AtomicU32,
-    message: String,
+    commit_message: String,
 }
 
-#[pymethods]
 impl PyWriteTransaction {
-    /// This is for testing
-    pub fn set_do_not_commit(&mut self, do_not_commit: bool) -> PyResult<()> {
-        self.do_not_commit = do_not_commit;
-        Ok(())
-    }
-    /// This is for testing
-    pub fn set_error_on_commit(&mut self, error_on_commit: bool) -> PyResult<()> {
-        self.error_on_commit = error_on_commit;
-        Ok(())
+    async fn new(repo: Arc<XetRepo>, branch: &str, commit_message: &str) -> Result<Self> {
+        Ok(Self {
+            pwt: Some(WriteTransaction::new(&repo, branch, commit_message).await?),
+            repo,
+            branch: branch.to_string(),
+            commit_message: commit_message.to_string(),
+        })
     }
-    pub fn open_for_write(&mut self, path: &str) -> PyResult<PyWFile> {
-        if let Some(ref mut tr) = self.transaction {
-            self.new_files
-                .push(format!("{}/{path}", self.branch).to_string());
-            let writer = exec_async_anyhow!(tr.open_for_write(path).await);
-            self.pending_write_files.fetch_add(1, Ordering::SeqCst);
 
-            Ok(PyWFile { writer })
+    fn access_inner<'a>(&'a self) -> Result<&'a Arc<RwLock<WriteTransaction>>> {
+        if let Some(t) = self.pwt.as_ref() {
+            Ok(t)
         } else {
-            Err(PyRuntimeError::new_err("Transaction already committed"))
+            // This means we've called close() on the transaction, then tried to use it.
+            Err(anyhow!(
+                "Transaction operation attempted after transaction completed."
+            ))
         }
     }
-    pub fn set_ready(&mut self) -> PyResult<()> {
-        let val = self.pending_write_files.load(Ordering::SeqCst);
 
-        // We need to check here because if all write files finished before this
-        // then none of them sees the ready signal.
-        if val == 0 {
-            // All write files finished
-            self.commit()
-        } else {
-            // Some files are pending, signal them to commit when they finish
-            self.ready_to_commit = true;
-            Ok(())
+    async fn complete_impl(&mut self, commit: bool, cleanup_immediately: bool) -> Result<()> {
+        let Some(tr) = self.pwt.take() else {
```

Comment:
It is possible that write files w.r.t this transaction finish before calling `end_transaction`, `complete`, thus this should be a NOP instead of error.

Reply:
Changed to info level and return Ok(())